File: 프로그래머스/lv2/17684. ［3차］ 압축/［3차］ 압축.py
def solution(msg):
    answer = []
    findDict = {}

    for idx, val in enumerate(range(ord("A"), ord("Z")+1), start=1):
        findDict[chr(val)] = idx

    start = 0
    end = len(msg)

    while True:
        a = msg[start:end]
        if a in findDict:
            answer.append(findDict[a])
            if end >= len(msg):
                return answer
            else:
                findDict[a+msg[end]] = len(findDict) + 1
                start += len(a)
                end = len(msg)
        else:
            end -= 1

File: 프로그래머스/lv2/17684. ［3차］ 압축/［3차］ 압축.js
function solution(msg) {
    var answer = [];
    const alpha = new Map();

    for (let i = "A".charCodeAt(0); i <= "Z".charCodeAt(0); i += 1) {
        alpha[String.fromCharCode(i)] = i - ("A".charCodeAt(0) - 1)
    }

    let start = 0;
    let end = start + 1;
    while (msg.length) {
        let x = msg.slice(start, end)
        if (!(x in alpha)) {
            alpha[x] = Object.keys(alpha).length + 1
            answer.push(alpha[msg.slice(start, end-1)])
            msg = msg.slice(end-1, msg.length)
            start = 0;
            end = start + 1;
            continue;
        }
        if (end == msg.length) {
            answer.push(alpha[x])
            break;
        }
        end += 1
    }

    return answer;
}